overlord/snapstate/storehelpers.go

```diff
@@ -224,14 +233,50 @@ func installInfo(ctx context.Context, st *state.State, name string, revOpts *Rev
 		InstanceName: name,
 	}
 
+	// TODO: support ignore-validation
+	enforcedSets, err := EnforcedValidationSets(st)
+	if err != nil {
+		return store.SnapActionResult{}, err
+	}
+
+	var requiredRevision snap.Revision
+	var requiredValSets []string
+	if enforcedSets != nil {
+		// check for invalid presence first to have a list of sets where it's invalid
+		invalidForValSets, err := enforcedSets.CheckPresenceInvalid(naming.Snap(name))
+		if err != nil {
+			if _, ok := err.(*snapasserts.PresenceConstraintError); !ok {
+				return store.SnapActionResult{}, err
+			} // else presence is optional or required, carry on
+		}
+		if len(invalidForValSets) > 0 {
+			return store.SnapActionResult{}, fmt.Errorf("cannot install snap %q due to enforcing rules of validation set %s", name, strings.Join(invalidForValSets, ","))
+		}
+		requiredValSets, requiredRevision, err = enforcedSets.CheckPresenceRequired(naming.Snap(name))
+		if err != nil {
+			return store.SnapActionResult{}, err
+		}
+	}
+
+	if len(requiredValSets) > 0 {
+		setActionValidationSets(action, requiredValSets)
+		if !requiredRevision.Unset() {
+			action.Revision = requiredRevision
+		}
+	}
+
+	// check if desired revision matches the revision required by validation sets
+	if !requiredRevision.Unset() && !revOpts.Revision.Unset() && revOpts.Revision.N != requiredRevision.N {
+		return store.SnapActionResult{}, fmt.Errorf("cannot install snap %q at requested revision %s, revision %s required by validation sets: %s", name, revOpts.Revision, requiredRevision, strings.Join(requiredValSets, ","))
```

Comment:
this error should mention --ignore-validation once we support that here, because specifying --revision and --ignore-validation should then work